### ISSUE TITLE
indexeddb adapter can break when opened in multiple tabs

### ISSUE THREAD

**Issue by srova** [pouchdb/pouchdb#8209]
### Issue
We have an application using pouchDB to make offline functionality possible, in live replication with a remote CouchDB cluster.
We recently moved from idb adapter to indexeddb adapter in order to perform better with native IdB indexes.
The app performance are now ok, but we have some cases where the end users experiment occasionally the following error:

Database has a global failure

launched by the new indexededb adapter in the idb.onabort callback.
When this happens, the download replication stream crashes, instead the upload one and the app local data browsing still work.
In all cases we had to start a fresh download of all local data from server to overcome this issue
Any ideas of possible causes or what to check?
Thanks

### Info
- Environment: (Node.js/browser/hybrid app.)
- Platform: (Electron app)
- Adapter: (IndexedDB)
- Server: (CouchDB)

**Comment by srova** [comment]
An other addition info: the crash described above is preceeded by the following error, launched inside pouchdb.min.js:

Cannot read property filter of undefined

Unfortunatly since the js is minified and we are in production enviroment we are not able to retrieve any forther detail right now

**Comment by elaynelemos** [comment]
@geidsonc

**Comment by srova** [comment]
For anyone interested we were finaly able to figure out what was happening:
- our remote CouchDB database contains many attachments
- when running in the electron app, sometimes (rarely - but our application is used every day by dozens of end users), an attachment download fails, and a txn.abort() is called (we still have not understood why it fails writing locally the attachment)
- this causes that the onAbort callback inside **pouchdb-indexeddb** is called, but unfortunatly the error variable is null, so the callback seems to be invoked for a successful result, even if the results array is undefined 


```
preProcessAttachments().then(function () {
    api._openTransactionSafely([DOC_STORE, META_STORE], 'readwrite', function (err, _txn) {
....
      txn.onabort = function () {
        callback(error);
      };
....

```

- this causes an uncatched exeption in **pouchdb-core, at adapter.js line 887** (trying to call .filter on res, that is undefined, causing the downstream syncronization to freeze

```
  return this._bulkDocs(req, opts, function (err, res) {
    if (err) {
      return callback(err);
    }
    if (!opts.new_edits) {
      // this is what couch does when new_edits is false
      res = res.filter(function (x) {
        return x.error;
      });
    }
    // add ids for error/conflict responses (not required for CouchDB)
    if (!isRemote(adapter)) {
      for (var i = 0, l = res.length; i < l; i++) {
        res[i].id = res[i].id || ids[i];
      }
    }

    callback(null, res);
  });
```

**Comment by msotnikov** [comment]
We have same issue. 
When tab #1 is syncronizing, and then start sync on tab #2 then first tab is fall with error.
We don't have any atachments in database.

Environment: browser, rxdb library
Platform: windows
Adapter: IndexedDB
Server: CouchDB

**Comment by msotnikov** [comment]
Are there maintainers here?

**Comment by srova** [comment]
@msotnikov I can confirm that even in our application the issue is not strictly related to attachments, even if with attachment it happens more frequently. Did you find that is something related with councurrency?

I would appreciate if someone of pouchdb mantainers cloud give us an hand

**Comment by msotnikov** [comment]
We spent a lot of days to find the reason and now I understand why the adapter fails.
tldr: New plugin is experimental (see below) and don't support several tabs.

1. Each new query create new connection (read or write). Not new tab!
2. New connection bump up the db version (depend on new Date().valueOf() ) and therefore other tabs receive .onVersionChange event and close connection.
3. For each db ("table") plugin creating a seq index with increment logic on add rows. Simultaneously inserting will fail. Code has coincident TODO
https://github.com/pouchdb/pouchdb/blob/1ba879709fdf6306dc3f8e9d2d67f4060ade6e50/packages/node_modules/pouchdb-adapter-indexeddb/src/bulkDocs.js#L214
4. Current fallback algoritm expect only one db and don't check any locks except version changes.

**Comment by srova** [comment]
@msotnikov Thanks for sharing your analysis.

Actually our application works in live sync with the server database, so, even if we have just one tab opened (we use electron, and we have just one instance of the app running), the simultaneous insert scenario is possible if the user saves data locally while a burst of insert/updates is coming from the server. 

If we have attachments involved in the download sync, this is obviously slower (before every doc write pouch downloads the file): so what happens if the user saves a record while a collection of attahment is downloaded from the server?

We have also experimented that in some cases the local documents writes started to fail systematically, even restarting the app, even if the user is offline. We have overcome this issue by doing a fresh download of server data, but we suspect that, for the reasons above, the seq field in the meta object store has become lower than the max seq in the doc object store. Has anyone had this problem?

Thanks

**Comment by msotnikov** [comment]
We have some success with replacing integer values by cuid (https://github.com/ericelliott/cuid)

No more conflicts with `seq` duplicates. But it should be tested. Probably there may be 2 problems (at least): not all data will be sync from local to remote database, not all docs will be present in changes event. For us it doesn't matter. Please share you experiense if you will try.

https://github.com/pouchdb/pouchdb/blob/1ba879709fdf6306dc3f8e9d2d67f4060ade6e50/packages/node_modules/pouchdb-adapter-indexeddb/src/bulkDocs.js#L214

replace with 
```
doc.seq = cuid();
```

and 

https://github.com/pouchdb/pouchdb/blob/1ba879709fdf6306dc3f8e9d2d67f4060ade6e50/packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js#L170

with
```
metadata.seq = cuid();
```

**Comment by srova** [comment]
Just an update about this issue.
By increasing our app logging we had the proof that the problem is related with race conditions in bulkDocs, and that the update_seq field in the "meta" objectstore is sometimes saved with a value below the max seq in the docs objectstore.

As a workaround, now we check that if meta.update_seq is < MAX(docs.seq) then we save meta with meta.update_seq = MAX(docs.seq) + 1. But it isjust a workaround, it does not solve the issue.

Unfortunatly we cannot use @msotnikov fix, because we need seq to be an increasing integer to preserve bidirectional synch (all data in the client must be in the server).

In the previous idb adapter the seq was an autoincrement in a separate objectstore, so it was thread safe. To have the same behaviour we need to put an asynch call instead of:
```
 doc.seq = ++metadata.seq; 
```
to have a thread safe incremental number, but the modification is not so trivial.

Is there still any mantainer for this plugin?

**Comment by msotnikov** [comment]
> Unfortunatly we cannot use @msotnikov fix, because we need seq to be an increasing integer to preserve bidirectional synch (all data in the client must be in the server).

Why doesn't it work? You can use another field for unique id / for sync. The sequence required only for bulk reads. I suppose. Did you check it?

**Comment by srova** [comment]
@msotnikov Actually I didn't tried it, because of what you said about the 2 problems.. :-)  I will try it and do the check (bulk reads).
I'll let you know as soon as possible

**Comment by daleharvey** [comment]
Thanks for the investigation @msotnikov, that is something we dont have good test coverage for

**Comment by daleharvey** [comment]
There seems to be possibly multiple bugs being discussed here as @msotnikov has identified one I will make this issue reference that, if you are having other issues please open a seperate bug, cheers

**Comment by rlamarche** [comment]
Hi, I think that I have encountered this bug two times. I was not using multiple times.

The error I get is :


```
TypeError: Cannot read property '<property name>' of undefined
    at getKeyFromDoc (index-browser.es.js:431)
    at filterInclusiveStart (index-browser.es.js:446)
    at index-browser.es.js:1298
    at ZoneDelegate.invoke (zone-evergreen.js:364)
    at Object.onInvoke (core.js:28205)
    at ZoneDelegate.invoke (zone-evergreen.js:363)
    at Zone.run (zone-evergreen.js:123)
    at zone-evergreen.js:857
    at ZoneDelegate.invokeTask (zone-evergreen.js:399)
    at Object.onInvokeTask (core.js:28193)
```

I'm on chrome using the new indexed db adapter and pouchdb / pouchdb-find at version 7.2.2
(pouchdb-browser & pouchdb-find at 7.2.2 in package.json)

Here my complete method that's fail after many calls (not the first) :

```
  async deleteElement(uuid: string) {

    // First we remove the file associated with the element
    let doc = await this.db.get(uuid);
    if (doc.downloaded) {
      try {
        const docFile = await this.fileDb.get(uuid);
        await this.fileDb.remove(docFile);
        console.log('ellement file removed: ', uuid);
      } catch (error) {
        console.log('unable to remove file associated with element, ignoring: ' + uuid, error);
      }
    }

    console.log('reload doc: ', uuid, doc);
    // We reload the doc to prevent any conflicts
    doc = await this.db.get(uuid);
    console.log('remove doc: ', uuid, doc);
    await this.db.remove(doc); // This call generates the error after 10 successful, and 1 more a few times after (passing through ngrx)
  }
```

When I remove the line `doc = await this.db.get(uuid);` just before the remove call, the bug won't happen anymore.
My hypothesis is that when I reload the document, the previous one is "freed" by the garbage collector, so appear to be undefined for subsequent asynchronous calls to build the index.

I've already seen this bug in another part of my app, an object was undefined but regarding the code it wasn't possible.

I hope it can help.

Note that my database has > 1000 documents and some indexes.

**Comment by msotnikov** [comment]
Oh no

**Comment by srova** [comment]
Still open?

**Comment by quolpr** [comment]
The issue is still here

**Comment by quolpr** [comment]
@srova could you remove stale label & reopen it?

**Comment by srova** [comment]
Unfortunately I cannot :-(
I think just mantainers can do this

**Comment by quolpr** [comment]
@msotnikov I can confirm that sync with `cuid` works incorrectly. In my case, it loads all changes from remote DB on each app start.

**Comment by AdrianoFerrari** [comment]
Can we get some maintainer eyes on this? @daleharvey ?

My users cannot use my writing app in multiple tabs, which they often want to do (one document per tab).
That's a pretty severe limitation for them.

I can try digging through the code on setting seq values, but I admit I'd be groping around in the dark. What can I do to help?

**Comment by AdrianoFerrari** [comment]
I've found a solution that seems to work, at least in my case.

Turns out that the issue is that http1 connections get stalled if we have too many going to the same domain.

Found this on RxDB website (https://rxdb.info/replication-couchdb.html):

>Since CouchDB only allows synchronization through HTTP1.1 long polling requests there is a limitation of 6 active synchronization connections before the browser prevents sending any further request. This limitation is at the level of browser per tab per domain (some browser, especially older ones, might have a different limit, see here).

> Since this limitation is at the **browser** level there are several solutions:

> 1. Use a proxy (ex: HAProxy) between the browser and CouchDB and configure it to use HTTP2.0, since HTTP2.0 doesn't have this limitation (RECOMMENDED)
> 2. Use only a single database for all entities and set a "type" field for each of the documents
> 3. Create multiple subdomains for CouchDB and use a max of 6 active synchronizations (or less) for each

I changed my nginx config from:
```
server {
   listen [::]:443 ssl;
   listen 443 ssl;
   # etc...
}
```
to
```
server {
   listen [::]:443 ssl http2;
   listen 443 ssl http2;
   # etc...
}
```

And now multi tab PouchDB/CouchDB works like a charm!

**Comment by dheimoz** [comment]
Thank you very much for sharing the solution @AdrianoFerrari 
This is the best way to contribute to this project.

Cheers.

**Comment by quolpr** [comment]
@AdrianoFerrari, it seems to be another issue. But I was debugging this issue, and I discovered the race condition as guys describe before. The issue is still here and still happens occasionally. But on a large number of writes, it may happen more often

**Comment by msotnikov** [comment]
@AdrianoFerrari @quolpr @daleharvey @dheimoz It is definitely another issue. Http2 helps fix staled connections. Run without http2 can speed up the election problems, but it is not the reason.  The issue must be reopened.

**Comment by quolpr** [comment]
Yep, totally agree. And @msotnikov pointed where race condition happens

**Comment by dheimoz** [comment]
@msotnikov and @quolpr I wonder if you have find the way to replicate the issue.

**Comment by msotnikov** [comment]
> I wonder if you have find the way to replicate the issue.

@dheimoz Maintainers should first disable the stale bot.

**Comment by dheimoz** [comment]
Hello @msotnikov , unfortunately it is not my call to disable the stale bot. But you can freely ping me should we need this issue to remain open.

**Comment by msotnikov** [comment]
https://github.com/pouchdb/pouchdb/issues/8494

**Comment by Nuu-maan** [comment]
Hey everyone 👋

I’d like to work on this issue and see if I can make a proper fix for the race condition in the `indexeddb` adapter that breaks sync in multi-tab scenarios. Based on the discussion, I’ll start by trying to reproduce the failure with parallel writes and see how the `seq` mismatch occurs. Then I’ll test if there’s a clean way to implement a safer, atomic increment strategy like the one used in the older idb adapter (possibly via a separate object store).

Let me know if anyone else is actively working on this. Otherwise, I’ll begin investigating and send a PR soon.
